File: 687-longest-univalue-path/687-longest-univalue-path.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestUnivaluePath = function (root) {
  let maxPath = 0;

  const helper = (node, val = -1) => {
    if (!node) return 0;

    const left = helper(node.left, node.val);
    const right = helper(node.right, node.val);
    maxPath = Math.max(maxPath, left + right);

    return node.val === val ? Math.max(left, right) + 1 : 0;
  };

  helper(root);

  return maxPath;
};

File: 687-longest-univalue-path/687-longest-univalue-path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        self.path = 0
        def helper(node, val=-1):
            if not node:
                return 0
            left = helper(node.left, node.val)
            right = helper(node.right, node.val)
            self.path = max(self.path, left + right)
            return max(left, right) + 1 if node.val == val else 0
          
        
        helper(root)
        return self.path